### PR TITLE
Handle complex JSON payloads for Call Choreo function and call Analytics function

### PR DESCRIPTION
**Pull request by shanggeeth**
$Subject

### Depends on
- https://github.com/wso2/carbon-identity-framework/pull/5867

### CODE REVIEW

**Review by ZiyamSanthosh — DISMISSED** [review]

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 1]
```diff
@@ -213,4 +215,31 @@ public void cancelled() {
         });
         JsGraphBuilder.addLongWaitProcess(asyncProcess, eventHandlers);
     }
+
+    private Map<String, Object> getPayloadDataMap(Map<String, Object> payloadData) {
+        Map<String, Object> payloadDataMap = new HashMap<>();
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:220]: missing new line

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 2]
```diff
@@ -213,4 +215,31 @@ public void cancelled() {
...
+        for (Map.Entry<String, Object> entry : payloadData.entrySet()) {
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:221]: is this null safe? can `payloadData` be null?

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 3]
```diff
@@ -213,4 +215,31 @@ public void cancelled() {
...
+                payloadDataMap.put(entry.getKey(), processList((List<Object>) value));
+            } else {
+                payloadDataMap.put(entry.getKey(), value);
+            }
+        }
+        return payloadDataMap;
+    }
+
+    private List<Object> processList(List<Object> list) {
+        List<Object> resultList = new ArrayList<>();
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:235]: Missing new line

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 4]
```diff
@@ -213,4 +215,31 @@ public void cancelled() {
...
+        for (Object item : list) {
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:236]: Can `list` be null?

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java [hunk 5]
```diff
@@ -213,4 +215,31 @@ public void cancelled() {
...
+            if (item instanceof Map) {
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:237]: can `item` be a List also?

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 1 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:224]: Added with ecb4acd9b9e89de443c225c4f58df10206068ab1

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 2 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:225]: Better to handle NPE here. Addressed with ecb4acd9b9e89de443c225c4f58df10206068ab1

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 3 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:240]: Added with ecb4acd9b9e89de443c225c4f58df10206068ab1

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 4 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:241]: List cannot be null here

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java (hunk 5 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/analytics/CallAnalyticsFunctionImpl.java:242]: This is possible. Handled with ecb4acd9b9e89de443c225c4f58df10206068ab1

**Comment by janakamarasena** [comment]
lets add the necessary unit tests for this scenario along with this PR

File: components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/analytics/analytics-payload.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        "arrayOfObjectsBooleanKey": true
+      },
+      {
+        "arrayOfObjectsStringKey": "arrayOfObjectsStringValue2",
+        "arrayOfObjectsNumberKey": 161718,
+        "arrayOfObjectsBooleanKey": false
+      }
+    ]
+  }
+}
```
Comment (ImalshaG) [inline components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/analytics/analytics-payload.json:39]: Add a new line

Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.analytics/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/analytics/analytics-payload.json]: Addressed with b708cb6175c7f325dfae500c417612a0385aa14d

File: components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/choreo-payload.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
+{
+  "stringKey": "stringValue",
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/choreo-payload.json:2]: Don't we have to consider null values being sent in the payload?

File: components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/payload-test-sp.xml [hunk 8]
```diff
@@ -0,0 +1,133 @@
+<!--
+  ~ Copyright (c) 22024, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/payload-test-sp.xml:2]: Check license year

File: components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java [hunk 9]
```diff
@@ -247,6 +263,26 @@ public ServiceProvider loadServiceProviderFromResource(String spFileName, Object
         return ServiceProvider.build(documentElement);
     }
 
+    public JsonObject loadJson(String jsonFileName, Object loader) throws JsTestException {
+        try (InputStream inputStream = loader.getClass().getResourceAsStream(jsonFileName)) {
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java:267]: Add new line. Check similar scenarios

File: components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java [hunk 10]
```diff
@@ -247,6 +263,26 @@ public ServiceProvider loadServiceProviderFromResource(String spFileName, Object
...
+                } else {
+                    throw new JsTestException("The JSON file does not contain a valid JSON object: " + jsonFileName);
+                }
+            }
+        } catch (IOException | JsonSyntaxException e) {
+            throw new JsTestException("Error in reading or parsing JSON file at: " + jsonFileName, e);
+        }
+    }
+
+
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java:285]: Remove additional lines. Check other placs too

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-get-response.json [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+        "arrayOfObjectsBooleanKey": true
+      },
+      {
+        "arrayOfObjectsStringKey": "arrayOfObjectsStringValue2",
+        "arrayOfObjectsNumberKey": 161718,
+        "arrayOfObjectsBooleanKey": false
+      }
+    ]
+  }
+}
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-get-response.json:39]: Add new line

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java [hunk 12]
```diff
@@ -475,4 +506,32 @@ public Map<String, String> dummyTokenEndpoint(@HeaderParam("Authorization") Stri
...
+                if (response != null) {
+                    JsonObject expectedResponse = sequenceHandlerRunner.loadJson(HTTP_GET_RESPONSE, this);
+                    Gson gson = new Gson();
+                    String dataStr = gson.toJson(response.getWrapped());
+                    JsonObject actualResponse = gson.fromJson(dataStr, JsonObject.class);
+                    return actualResponse.equals(expectedResponse);
+                }
+                return false;
+            } catch (JsTestException e) {
+                throw new RuntimeException(e);
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java:533]: Check on throwing runtime exception

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java [hunk 13]
```diff
@@ -475,4 +506,32 @@ public Map<String, String> dummyTokenEndpoint(@HeaderParam("Authorization") Stri
         }
         return response;
     }
+
+    @FunctionalInterface
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java:510]: Add comments explaining the use of the interface and its implementation

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImplTest.java [hunk 14]
```diff
@@ -491,4 +538,32 @@ public Map<String, String> dummyTokenEndpoint(@HeaderParam("Authorization") Stri
         }
         return response;
     }
+
+    @FunctionalInterface
+    public interface ResponseValidator {
```
Comment (ZiyamSanthosh) [inline components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImplTest.java:543]: Seems like this interface and its impl are duplicated in each test classes. Let's write them outside the test classes separately.

File: components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/choreo-payload.json (hunk 7 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/choreo-payload.json:2]: There's a limitation where null values are skipped when response is received and payload is sent which is not caused by any adaptive authentication mechanisms. Hence `Null` is skipped from the payload.
However the issue needs to identified and fixed.

File: components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/payload-test-sp.xml (hunk 8 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/choreo/payload-test-sp.xml]: Addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-get-response.json (hunk 11 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/resources/org/wso2/carbon/identity/conditional/auth/functions/http/http-get-response.json]: Addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java (hunk 12 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java]: We don't need to throw RuntimeException. Hence addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImplTest.java (hunk 14 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPPostFunctionImplTest.java]: We can extract the Interface. Addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

File: components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java (hunk 13 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.http/src/test/java/org/wso2/carbon/identity/conditional/auth/functions/http/HTTPGetFunctionImplTest.java]: Addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

File: components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java (hunk 9 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.conditional.auth.functions.test.utils/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/test/utils/sequence/JsSequenceHandlerRunner.java]: Addressed with 46c8e19185f03a9c7a8e20a87f85cae4e5e0ea96

**Review by ZiyamSanthosh — APPROVED** [review]

**Pull request merged**